[wip] gossiper basic functions

wfystx opened the pull request:
- HeartbeatReceiver function is used as a server, should be run in a go routine, keep listening income heartbeat
- SendHeartBeat function is used to send heartbeat in a time interval

I'm still struggling how to build the cli to better test these functions

wfystx committed:
basic functions

Bowenislandsong commented on gossiper/gossiper.go:
```diff
@@ -0,0 +1,21 @@
+
+func (g *gossiper) GetPeerNames() {
+	fmt.Println(g.peers)
+}
```
Always have an extra line.

Bowenislandsong commented on gossiper/gossiper.go:
```diff
@@ -0,0 +1,21 @@
+	delete(g.peers, name)
+}
+
+func (g *gossiper) GetPeerNames() {
```
```suggestion
func (g *gossiper) PrintPeerNames() {
```

Bowenislandsong commented on gossiper/gossiper.go:
```diff
@@ -0,0 +1,21 @@
+type gossiper struct {
+	name	string
+	ip		string
+	peers	map[string]gossiper
```
Might want a flat map? IMO, we do not need the entire network information. Wdyt?

wfystx replied:
Yeah we don't need the entire network information, I was thinking each node uses this map to keep its peers

Bowenislandsong commented on gossiper/gossiper_test.go:
```diff
@@ -0,0 +1,16 @@
+	g.name = "test"
+	g.AddPeer(gossiper{"peer", "localhost", make(map[string]gossiper)})
+	if _, ok := g.peers["peer"]; !ok {
+		t.Errorf("Add Failed")
```
Might wanna look at testing package -> https://godoc.org/github.com/stretchr/testify/assert and https://godoc.org/github.com/stretchr/testify/require

wfystx replied:
will do

Bowenislandsong reviewed (changes requested):
Good initial work. A better way maybe to draw out what you want to implement as a plan. Else this could go for a bit long.

Bowenislandsong commented:
Your title [wip] is enough to say it is work in progress. Your actual commit message should include the motivation of the work, what the implementation does, and why it would fit the profile.

Bowenislandsong commented on gossiper/gossiper.go:
```diff
@@ -0,0 +1,21 @@
+package gossiper
```
pkg/gossiper/peers.go

wfystx committed:
Update gossiper/gossiper.go

Co-authored-by: Bowen Song <bowenislandsong@gmail.com>

wfystx committed:
heartbeat

Bowenislandsong commented on go.mod:
```diff
@@ -0,0 +1,9 @@
+
+go 1.14
+
+replace github.com/Never-M/MyGossip => /Users/wangfuyao/go/src/github.com/Never-M/MyGossip
```
Why is this necessary?

wfystx replied:
If I don't do this, it would cause a package not found error. I haven't found a solution for it

Bowenislandsong commented on go.mod:
```diff
@@ -0,0 +1,9 @@
+replace github.com/Never-M/MyGossip => /Users/wangfuyao/go/src/github.com/Never-M/MyGossip
+
+require (
+	github.com/Never-M/MyGossip v0.0.0-00010101000000-000000000000 // indirect
```
Also don’t think this is necessary

Bowenislandsong replied:
Could just let it generate itself with go mod init

Bowenislandsong commented on pkg/gossiper/peers.go:
```diff
@@ -0,0 +1,86 @@
+	ctx = context.WithValue(ctx, "key", "value")
+	req, err := http.NewRequestWithContext(ctx, "GET", "http://localhost:8085/heartbeat", nil)
+	if err != nil {
+		log.Println(err)
```
Use log.info() from logrus https://github.com/sirupsen/logrus#example

wfystx replied:
Yes that's what I thought too. @Never-M can you add logrus to our project?

Bowenislandsong commented on pkg/gossiper/peers.go:
```diff
@@ -0,0 +1,86 @@
+	currentTime		time.Time
+}
+
+func NewGossiper(self string, address string) *gossiper {
```
(self, address string)

Also could just name self `name`

Bowenislandsong commented on pkg/gossiper/peers.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func NewGossiper(self string, address string) *gossiper {
+	g := &gossiper{
```
Return &gossiper{...

Bowenislandsong commented on pkg/gossiper/peers.go:
```diff
@@ -0,0 +1,86 @@
+		peers:make(map[string]gossiper),
+		client:&http.Client{},
+	}
+	return g;
```
Don’t need `;`

Bowenislandsong commented on pkg/gossiper/peers.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (g *gossiper) AddPeer(peer gossiper)  {
+	g.peers[peer.name] = peer
```
Might wanna check for dup

Bowenislandsong commented on pkg/gossiper/peers.go:
```diff
@@ -0,0 +1,86 @@
+	mux.HandleFunc("/heartbeat", g.HeartBeatHandler)
+	contextedMux := g.HeartBeatContext(mux)
+	log.Println("Start server on port :8085")
+	http.ListenAndServe(":8085", contextedMux)
```
Should make this at least a variable if not a global var

Bowenislandsong commented on pkg/gossiper/peers_test.go:
```diff
@@ -0,0 +1,16 @@
+	}
+	g.RemovePeer("peer")
+	if _, ok := g.peers["peer"]; ok {
+		t.Errorf("Remove Failed")
```
Try assert package from testify and it would be like Assert.False(t,ok)

Bowenislandsong commented on pkg/gossiper/peers_test.go:
```diff
@@ -0,0 +1,16 @@
+import "testing"
+
+func TestAddRemovePeer(t *testing.T)  {
+	g := new(gossiper)
```
Shouldn’t this be NewGossiper(“test”,...)?

Bowenislandsong reviewed (changes requested):
Some initial comments. This is kinda hard to do on an iPad with the github app. I do not recommend using the github app, not yet. It can’t code or review, I don’t know what it’s good for. 🤣🤣🤣

wfystx committed:
Added heartbeat test. Added result code types

wfystx committed:
Added heartbeat test. Added result code types

wfystx committed:
deleted main.go

wfystx committed:
leveldb init

Never-M committed:
db basic functions

Never-M committed:
logrus

wfystx committed:
pair programming

Bowenislandsong committed:
small fixes

After this fix, we will convert to use body part of the HTTP to transfer data instead of ctx. Context should just be used for signalling

wfystx committed:
pair programming 2

wfystx committed:
merge

Never-M committed:
finish http request, sending body

Never-M committed:
change log.go newlog function

Never-M committed:
change log

wfystx commented on pkg/gossiper/db.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+//how to use return &mydb{} method？
+func newdb(path string) (int, *mydb) {
```
Capital

wfystx commented on pkg/gossiper/db.go:
```diff
@@ -0,0 +1,70 @@
+	return SUCCEED, newdb
+}
+
+func (d *mydb) put(key, val string) int {
```
Capital
